apps/prairielearn/elements/pl-drawing/pl-drawing.py

```diff
@@ -494,3 +496,125 @@ def grade(element_html: str, data: pl.QuestionData) -> None:
         "weight": 1,
         "feedback": {"correct": (score == 1), "missing": {}, "matches": matches},
     }
+
+
+def test(element_html: str, data: pl.ElementTestData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    name = pl.get_string_attrib(
+        element, "answers-name", defaults.element_defaults["answers-name"]
+    )
+    gradable = pl.get_boolean_attrib(
+        element, "gradable", defaults.element_defaults["gradable"]
+    )
+    if not gradable:
+        return
+
+    # Get raw correct answer
+    a_tru = data["correct_answers"][name]
+    grid_size = pl.get_integer_attrib(
+        element, "grid-size", defaults.element_defaults["grid-size"]
+    )
+    tol = pl.get_float_attrib(element, "tol", grid_size / 2)
+    angtol = pl.get_float_attrib(
+        element, "angle-tol", defaults.element_defaults["angle-tol"]
+    )
+
+    result = data["test_type"]
+    if result == "correct" and len(a_tru) > 0:
+        data["raw_submitted_answers"][name] = json.dumps(a_tru)
+        data["partial_scores"][name] = {
+            "score": 1,
+            "weight": 1,
+            "feedback": {
+                "correct": True,
+                "missing": {},
+                "matches": {
+                    element["id"]: True
+                    for element in a_tru
+                    if elements.is_gradable(element["gradingName"])
+                    and element["graded"]
+                },
+            },
+        }
+
+    elif result == "incorrect" and len(a_tru) > 0:
+        data["raw_submitted_answers"][name] = copy.deepcopy(a_tru)
+        for i, element in enumerate(a_tru):
+            if (
+                not elements.is_gradable(element["gradingName"])
+                or not element["graded"]
+            ):
+                continue
+
+            mutated = False
+            if "x1" in element and "y1" in element:
```

Comment:
Can we express all this something more like this?

```python
for prop in ["x1", "y1", "x2", "y2", ...]:
    if prop in element:
        data["raw_submitted_answers"][name][i][prop] += (
            random.choice([1, -1]) * 1.1 * tol
        )
        mutated = True
```

Reply:
Sure. https://github.com/PrairieLearn/PrairieLearn/pull/11868/commits/116a6c14c903e61813a8bef20e840c1124cc3211

Reply:
For my own understanding, why does it matter that they're done in pairs?

Reply:
To avoid a situation where we give partial credit accidentally. I suppose it doesn't _really_ matter, but some of these elements are set up strangely, and I would prefer it gives an error if e.g. only `x1` is an attribute.

Reply:
Really I should be doing `x1, x2, y1, y2` as one attribute set and `x1, x2, y1, y2, x3, y3` as another attribute set to avoid partial credit.

Reply:
I think I'm missing something. How is partial credit possible at all here? Doesn't `grade(...)` on an element just return a boolean?

Reply:
You are right, apologies. In that case I was setting a bunch of values, all of which didn't impact grading.

Reply:
[db4d2e7](https://github.com/PrairieLearn/PrairieLearn/pull/11868/commits/db4d2e709d03f4f3b823bf5b3eb675d93895068a)